Add wrap function

BSteephenson opened the pull request:
Adds a line wrapping function that limits line length to a certain width
In response to issue #278 

Let me know if there's anything more I should do (more tests, documentation, etc). I'm glad to help out.

stoeffel commented:
I will have a look at it on the WE.Can you add the documentation to the `README.markdown`? Thanks for the PR.

stoeffel commented on README.markdown:
````diff
@@ -267,6 +267,15 @@ lines("Hello\nWorld");
 // => ["Hello", "World"]
 ```
 
+#### wrap(str, width, seperator, cut) => string
````
You should document the complete API here (defaults and some more examples).

stoeffel commented on test/strings.js:
```diff
@@ -899,5 +899,10 @@ $(document).ready(function() {
+    equal(_.wrap("My name is", 2, '.', false), "My .name .is", 'works with width 2 and cut = false');
+    equal(_.wrap("My name is", 2, '.', true), "My. n.am.e .is", 'works with width 2 and cut = true');
+    equal(_.wrap("My name is", 3, '.', false), "My .name .is", 'works with width 3 and cut = true');
+    equal(_.wrap("My name is", 3), "My \nname \nis", 'Default parameters work');
```
You should add more tests.
* edgecases
* defaults

Checkout the other tests.

stoeffel commented on wrap.js:
```diff
@@ -0,0 +1,55 @@
+
+
+module.exports = function wrap(str, width, seperator, cut){
+	str = str || "";
```
You should use [`makeStr`](https://github.com/epeli/underscore.string/blob/master/helper/makeString.js) here.

stoeffel commented on wrap.js:
```diff
@@ -0,0 +1,55 @@
+// str       : defualt ""    :  input line
+// width     : defualt 75    :  width of column of text
+// seperator : defualt '\n'  :  character that seperates lines
+// cut       : defualt false :  should it start a new line in the middle of a word?
```
You can move the jsdoc to the readme. We don't use jsdoc at the moment.

stoeffel commented:
Can you rebase?

BSteephenson commented:
Rebased with a mocha test
Let me know if there's anything else I should do =)

stoeffel commented:
If you add this test:
`equal(wrap("My name is", -1), "My name is", "Just return original line if width <= 0")`
You have 100% coverage.

Otherwise it looks good to me.

BSteephenson commented:
Done and squashed.

stoeffel commented on wrap.js:
```diff
@@ -0,0 +1,55 @@
+		return str;
+	}
+	else if(!cut){
+		words = str.split(" ");
```
How about we use `underscore.string/words' here?

stoeffel commented on wrap.js:
```diff
@@ -0,0 +1,55 @@
+	else if(!cut){
+		words = str.split(" ");
+		for(var i = 0; i < words.length - 1; i++){
+			//preserve the spaces for all words except the last one
```
I'm actually not sure if we should preserve the space. This behaviour is somehow different than I would expect it to be.

BSteephenson commented:
Let's talk about the specifics of what this is supposed to do.

wrap('this is a sentence', 4, \n, false)

What should this return? Right now it would return 'this \nis a \nsentence' which I'm realizing now doesn't make much sense. Would 'this\nis a\nsentence' be more obvious?

stoeffel commented:
> Would 'this\nis a\nsentence' be more obvious?

Yes, I think so.

PHP's `wordwrap` also drops spaces at the end. http://us3.php.net/manual/en/function.wordwrap.php
We could add an option for this, but it already has 4 arguments...

epeli commented:
Options object maybe. Could be done as backwards compatible with type checking.

stoeffel commented:
:+1: for an options object I could think of some other options we could add in the future f.e. dontBreakHTMLTags

BSteephenson commented:
Now wrap('this is a sentence', {width: 4, seperator: '\n', cut: false}) returns 'this\nis a\nsentence'

stoeffel commented:
Options object looks good. Could you add an option for dropping trailing spaces as discussed above? And updat the readme? Thanks for your work

BSteephenson commented:
I added a trailingSpaces options which, if true, fills each line with trailing spaces to make them at least `width` long (I'm not sure if this is what you meant). I also updated the Readme.

stoeffel commented:
Sorry, I actually meant this:

> What should this return? Right now it would return 'this \nis a \nsentence' which I'm realizing now doesn't make much sense. Would 'this\nis a\nsentence' be more obvious?

Maybe call the option `preserveSpace`. Sry for the confusion.

stoeffel commented:
but the trailingSpaces option is nice too :+1:

stoeffel commented:
@BSteephenson Would be great if you would add the `preserveSpaces` (or take a better name) option so we can merge this.

Thanks

BSteephenson commented:
Working on this atm. I am having preserveSpaces have a higher precedence over trailingSpaces.

stoeffel commented on README.markdown:
```diff
@@ -269,6 +269,25 @@ lines("Hello\nWorld");
 
+#### wrap(str, options) => string
+
+Splits a line `str` (default '') into several lines of size `options.width` (default 75) using a `options.seperator` (default '\n'). If `options.trailingSpaces` is true, make each line at least `width` long using trailing spaces. If `options.cut` is true, create new lines in the middle of words. If `options.trailingSpaces` is true, preserve the space that should be there at the end of a line.
```
> `options.trailingSpaces` is true, preserve the space that should be there at the end of a line.

This should be `options.preserveSpaces`

stoeffel commented on README.markdown:
```diff
@@ -269,6 +269,25 @@ lines("Hello\nWorld");
+wrap("Hello World", { width:5, seperator:'.', cut:true, trailingSpaces: true })
+// => "Hello. Worl.d    "
+
+wrap("Hello World", { width:5, seperator:'.', cut:false, preserveSpaces: true })
```
I would drop `cut:false` here and on line 280 since the default is `false` anyway

stoeffel replied:
should we state that `preserveSpaces` has no influence if `cut:true`?

stoeffel commented:
Besides my comments in the readme, this looks good to me. Thx a lot

BSteephenson commented:
Corrected all the oops's in Readme
Let me know if you want me to squash all the commits into one

stoeffel commented:
Squashing would be nice. I want have time today to check it one more time. But I will review it again in the beginning of next week and then merge it.
Thanks for your work.

stoeffel commented:
Did check it again. Could you squash? I will merge it.

BSteephenson commented:
Done =)

stoeffel commented:
nice thanks a lot.

cahrens commented:
@stoeffel when "wrap" was added, it wasn't added to the "exports" function as a conflict with underscore. I know mixing in underscore.string is now discouraged, but unfortunately we have a lot of old code written with the mixin paradigm. I am having difficulty upgrading underscore.string from 2.x due to the wrap conflict.

stoeffel commented:
Hey @cahrens. Good catch. Would be awesome if you could open a PR to include `wrap` in the exports.
I want to merge #481 and #482 first, before I publish the next version (if you have time and willing to review these two PRs would be cool).

Thanks

cahrens commented:
Sure, I will looking into it this weekend. I would like to add a test that checks for conflicting functions between underscore and underscore.string. I haven't looked at your test framework yet, but hope it's possible to add a test-only dependency on underscore.

cahrens commented:
@stoeffel I see that underscore is already a development dependency. Can you point me to documentation on how to run the tests locally? I poked around but couldn't find anything.
